builtin/google/cloudrun/platform.go

```diff
@@ -392,27 +396,140 @@ func (p *Platform) Deploy(
+) error {
+	if isWorkspaceDestroy {
+		client := run.NewNamespacesServicesService(apiService)
+		st.Step("Deleting service: %s", state.ApiName)
+
+		_, err := client.Delete(d.apiName()).Context(ctx).Do()
+		return err
+	}
+	rn := state.RevisionName
+	client := run.NewNamespacesRevisionsService(apiService)
+	st.Step("Deleting deployment with revision: %s", rn)
+
+	// First check if it's currently serving traffic
+	r, err := client.Get(rn).Context(ctx).Do()
+	if err != nil {
+		return err
+	}
+	// the slice of status conditions returned by GCP stores the reason
+	// for the most recent status transition at index [1]. We check that
+	// the revision is serving by verifying the operating status is active
+	// https://cloud.google.com/run/docs/reference/rest/v1/namespaces.revisions
+	cs := r.Status.Conditions[1]
+	if cs.Status == "True" {
+		st.Step(
+			terminal.StatusWarn,
+			fmt.Sprintf("Cannot destroy deployment with revision %q, as revision is actively receiving traffic.", rn),
+		)
+		return nil
+	}
+
+	_, err = client.Delete(state.RevisionName).Context(ctx).Do()
+	return err
+}
+
+// Deploy deploys an image to Cloud Run.
+func (p *Platform) Deploy(
+	ctx context.Context,
+	log hclog.Logger,
+	src *component.Source,
+	img *docker.Image,
+	deployConfig *component.DeploymentConfig,
+	ui terminal.UI,
+) (*Deployment, error) {
+	// Start building our deployment since we use this information
+	deployment := &Deployment{
+		Resource: &Deployment_Resource{
+			Location: p.config.Location,
+			Project:  p.config.Project,
+			Name:     src.App,
+		},
+	}
+	id, err := component.Id()
+	if err != nil {
+		return nil, err
+	}
+	deployment.Id = id
+
+	// Validate that the Docker image is stored in a GCP registry
+	// It is not possible to deploy to Cloud Run using external container registries
+	err = validateImageName(img.Image)
+	if err != nil {
+		return nil, status.Error(codes.FailedPrecondition, err.Error())
+	}
+
+	// Get the Cloud Run Locations available for this project
+	// Cloud Run is only available in a limited number of Locations, this may be further restricted
+	// by the users access
+	pls, err := deployment.getLocationsForProject(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Aborted, err.Error())
+	}
+
+	// Validate that the Location specified for the deployment is available for the project
+	err = validateLocationAvailable(p.config.Location, pls)
+	if err != nil {
+		return nil, err
+	}
+
+	// We'll update the user in real time
+	st := ui.Status()
+	defer st.Close()
+
+	rm := p.resourceManager(log)
+	if err := rm.CreateAll(
+		ctx, deployment, log, st, pls,
+		deployConfig, img, src,
+	); err != nil {
+		return nil, err
+	}
+
+	// Store our resource state
+	deployment.ResourceState = rm.State()
+
+	// Get our service state
+	servState := rm.Resource(rmResourceServiceName).State().(*Resource_Service)
+	if servState == nil {
+		return nil, status.Errorf(codes.Internal,
+			"service state is nil, this should never happen")
+	}
```

Comment:
What is the purpose of this chunk?

Reply:
It was there previously; I assume it's one of those "this should ever happen but just in case it does, let's exit early with an explicit error"

Reply:
Oh wait. I lied. that's my fault!

Reply:
Yeah OK double-checked, and it's basically a check on "did we set the state on the resource manager, is it there." It's arguably not needed, but in the weird case that it _is_ needed, we don't get into a situation where the user expects state management to be there and it isn't.